### ISSUE TITLE
Forbid to have "meaningless blocks"

### ISSUE THREAD

**Issue by sobolevn** [wemake-services/wemake-python-styleguide#382]
# Rule request

## Thesis

There are some keywords that can not be a single statement inside a loop's body.
Because it won't make any sense. Consider this tiny example:

```python
for number in range(5):
     break
```

And things like `raise` and `return` can not be executed several times inside a loop.

## Reasoning

This is a best practice. Since it totally looks like an error. Or a hack to get a first value from an iterable.

**Comment by sobolevn** [comment]
The same is applied to `pass`

**Comment by sobolevn** [comment]
Related: #366 

We can create a single violation for these two cases.
However, the checking logic differs.

**Comment by sobolevn** [comment]
There might be other meaningless blocks in the future as well.

**Comment by sobolevn** [comment]
Nodes that make blocks useless:
- `with`: `break`, `continue`, `pass`
- `for`, `while`: `return`, `break`, `continue`, `pass`, `raise`
- `try`: `break`, `continue`, `pass`, `raise`

**Comment by kxepal** [comment]
Does this one counts?
```
def genfun():
  while False:
    yield
  return
```

**Comment by Snawoot** [comment]
> Because it won't make any sense.

Construction with break may have sense in conjunction with `for ... else` clause.

Construction with raise and return make sense when you need to consume iterator exactly once.

**Comment by sobolevn** [comment]
`...` is useless as well.

**Comment by sobolevn** [comment]
@kxepal nope, `yield` is never useless.

@Snawoot can you please provide a code example?

**Comment by Snawoot** [comment]
@sobolevn Something like this:

```python
def head_or_nothing(it):
""" Returns first value from iterable or None if it is empty """
    for val in it:
        return val
```

```python
def iter_empty(it):
""" Checks if there are still remaining values in iterator.
    Useful in conjunction with itertools.tee() """
    for _ in it:
        break
    else:
        return True
    return False
```

Another useful construction. Actually, it is simplified `consume()` from [itertools recipes](https://docs.python.org/3.7/library/itertools.html#itertools-recipes):

```python
def consume(it):
""" Consume iterator till the end. """
    for _ in it:
        pass
```

Also, `raise` keyword might be used as some sort of "goto" operator in order to leave current code immediately and jump to specific location of external blocks. Sometimes it does make sense.

**Comment by kxepal** [comment]
```
def head_or_nothing(it):
  return next(it, None)

def iter_empty(it):
  thing = next(it, sentinel)
  return thing is not sentinel
```

That's the usecase for the last one? Wouldn't be better to just `del it`?

**Comment by Snawoot** [comment]
@kxepal That's ok, but those examples are still valid.

Consuming iterators make sense when iteration code has side effects and we want them to finish.

**Comment by sobolevn** [comment]
@Snawoot thanks for taking your time and creating these examples.

I agree with you that these are valid cases.
I also agree with @kxepal that there are better and more pythonic ways to solve these cases.

And the purpose of this app is to *force* people to research other possible ways and write better code as a result. 

All in all, I am now even more convinced that this rule should be added as is.
